File: Top 100 Codes/06-Advanced Codes related to Arrays/04-Find the Union and Intersection of the two sorted arrays/FindtheUnionandIntersectionofthetwosortedarrays.cpp
#include <iostream>
using namespace std;

void findUnionAndIntersection(int arr1[], int arr2[], int n1, int n2) {
    int i = 0, j = 0;
    int unionArr[100], intersectionArr[100]; // assuming size is enough
    int uIndex = 0, iIndex = 0;

    for (; i < n1 && j < n2;) {
        if (arr1[i] < arr2[j]) {
            unionArr[uIndex++] = arr1[i++];
        } else if (arr2[j] < arr1[i]) {
            unionArr[uIndex++] = arr2[j++];
        } else {
            unionArr[uIndex++] = arr1[i]; // or arr2[j]
            intersectionArr[iIndex++] = arr1[i];
            i++;
            j++;
        }
    }

    // Remaining elements of arr1
    for (; i < n1; i++) {
        unionArr[uIndex++] = arr1[i];
    }

    // Remaining elements of arr2
    for (; j < n2; j++) {
        unionArr[uIndex++] = arr2[j];
    }

    // Print union
    cout << "Union: ";
    for (int k = 0; k < uIndex; k++) {
        // Skip duplicates
        if (k == 0 || unionArr[k] != unionArr[k - 1])
            cout << unionArr[k] << " ";
    }
    cout << endl;

    // Print intersection
    cout << "Intersection: ";
    for (int k = 0; k < iIndex; k++) {
        // Skip duplicates
        if (k == 0 || intersectionArr[k] != intersectionArr[k - 1])
            cout << intersectionArr[k] << " ";
    }
    cout << endl;
}

int main() {
    int arr1[] = {1, 2, 4, 5, 6};
    int arr2[] = {2, 3, 5, 7};
    int n1 = sizeof(arr1) / sizeof(arr1[0]);
    int n2 = sizeof(arr2) / sizeof(arr2[0]);

    findUnionAndIntersection(arr1, arr2, n1, n2);

    return 0;
}

File: Top 100 Codes/06-Advanced Codes related to Arrays/04-Find the Union and Intersection of the two sorted arrays/FindtheUnionandIntersectionofthetwosortedarrays.py
def find_union_and_intersection(arr1, arr2):
    n1 = len(arr1)
    n2 = len(arr2)
    i = j = 0
    union = []
    intersection = []

    # Traverse both arrays
    while i < n1 and j < n2:
        if arr1[i] < arr2[j]:
            union.append(arr1[i])
            i += 1
        elif arr2[j] < arr1[i]:
            union.append(arr2[j])
            j += 1
        else:
            union.append(arr1[i])  # or arr2[j]
            intersection.append(arr1[i])
            i += 1
            j += 1

    # Add remaining elements of arr1
    while i < n1:
        union.append(arr1[i])
        i += 1

    # Add remaining elements of arr2
    while j < n2:
        union.append(arr2[j])
        j += 1

    # Remove duplicates while preserving order
    def remove_duplicates(seq):
        result = []
        seen = set()
        for item in seq:
            if item not in seen:
                seen.add(item)
                result.append(item)
        return result

    union = remove_duplicates(union)
    intersection = remove_duplicates(intersection)

    print("Union:", union)
    print("Intersection:", intersection)

# Example
arr1 = [1, 2, 4, 5, 6]
arr2 = [2, 3, 5, 7]
find_union_and_intersection(arr1, arr2)
